flecsi/execution/hpx/execution_policy.h

```diff
@@ -28,183 +30,250 @@
     typename RETURN,
     typename ARG_TUPLE,
     typename... ARGS>
   static decltype(auto) execute_task(ARGS &&... args) {
+
     context_t & context_ = context_t::instance();
 
-    // Get the function and processor type.
-    auto fun = context_.task<KEY>();
+    auto function = context_.function(TASK);
+
+    using annotation = flecsi::utils::annotation;
+#if defined(ENABLE_CALIPER)
+    auto tname = context_.function_name(TASK);
+#else
+    /* using a placeholder so we do not have to maintain function_name_registry
+       when annotations are disabled. */
+    std::string tname{""};
+#endif
+
+    // Make a tuple from the task arguments.
+    utils::convert_tuple_t<ARG_TUPLE, std::decay_t> task_args =
+      std::make_tuple(std::forward<ARGS>(args)...);
+
+    annotation::begin<annotation::execute_task_prolog>(tname);
+    // run task_prolog to copy ghost cells.
+    task_prolog_t task_prolog;
+    task_prolog.walk(task_args);
+#if defined(FLECSI_USE_AGGCOMM)
+    task_prolog.launch_copies();
+    task_prolog.launch_sparse_copies();
+#endif
+    annotation::end<annotation::execute_task_prolog>();
+
+    annotation::begin<annotation::execute_task_user>(tname);
+
+    //     hpx_future_u<RETURN> future;
+    //     auto processor_type = context_.processor_type<TASK>();
+    //     if(processor_type == processor_type_t::mpi) {
+    //
+    //       {
+    //         clog_tag_guard(execution);
+    //         clog(info) << "Executing MPI task: " << TASK << std::endl;
+    //       }
+    //
+    //       future = executor_u<RETURN, ARG_TUPLE>::execute(
+    //         context_t::instance().get_mpi_executor(), std::move(function),
+    //         task_args);
+    //     }
+    //     else {
+    //       future = executor_u<RETURN, ARG_TUPLE>::execute(
+    //         context_t::instance().get_default_executor(),
+    //         std::move(function), task_args);
+    //     }
+    hpx_future_u<RETURN> future = executor_u<RETURN, ARG_TUPLE>::execute(
+      context_t::instance().get_default_executor(), std::move(function),
+      task_args);
+
+    annotation::end<annotation::execute_task_user>();
+
+    annotation::begin<annotation::execute_task_epilog>(tname);
+    task_epilog_t task_epilog;
+    task_epilog.walk(task_args);
+    annotation::end<annotation::execute_task_epilog>();
+
+    annotation::begin<annotation::execute_task_finalize>(tname);
+    finalize_handles_t finalize_handles;
+    finalize_handles.walk(task_args);
+    annotation::end<annotation::execute_task_finalize>();
+
+    constexpr size_t ZERO =
+      flecsi::utils::const_string_t{EXPAND_AND_STRINGIFY(0)}.hash();
+
+    if constexpr(REDUCTION != ZERO) {
 
-    auto processor_type = context_.processor_type<KEY>();
-    if(processor_type == processor_type_t::mpi) {
-      {
-        clog_tag_guard(execution);
-        clog(info) << "Executing MPI task: " << KEY << std::endl;
-      }
+      return future
+        .then([&](hpx_future_u<RETURN> && future) {
```

Comment:
Restrict capture to necessary values to avoid capturing things that might go out of scope.